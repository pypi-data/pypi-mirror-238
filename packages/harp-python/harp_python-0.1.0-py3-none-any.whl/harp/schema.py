from os import PathLike
from pathlib import Path
from typing import TextIO, Union
from harp.model import Model, Registers
from pydantic_yaml import parse_yaml_raw_as

_common_yaml_path = Path(__file__).absolute().parent.joinpath("common.yml")


def _read_common_registers(file: Union[str, PathLike, TextIO]) -> Registers:
    try:
        with open(file) as fileIO:
            return _read_common_registers(fileIO)
    except TypeError:
        return parse_yaml_raw_as(Registers, file.read())


def read_schema(
    file: Union[str, PathLike, TextIO], include_common_registers: bool = True
) -> Model:
    try:
        with open(file) as fileIO:
            return read_schema(fileIO, include_common_registers)
    except TypeError:
        schema = parse_yaml_raw_as(Model, file.read())
        if not "WhoAmI" in schema.registers and include_common_registers:
            common = _read_common_registers(_common_yaml_path)
            schema.registers = dict(common.registers, **schema.registers)
            schema.bitMasks = dict(common.bitMasks, **schema.bitMasks)
            schema.groupMasks = dict(common.groupMasks, **schema.groupMasks)
        return schema
