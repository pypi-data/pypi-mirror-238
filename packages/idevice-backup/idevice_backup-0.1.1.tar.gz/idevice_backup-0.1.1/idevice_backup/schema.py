from datetime import datetime
from pathlib import Path
from typing import List

from pydantic import BaseModel, Field, RootModel


class Snapshot(BaseModel):
    id: str
    short_id: str
    time: datetime
    tree: str
    paths: List[Path]
    hostname: str
    username: str
    uid: int
    gid: int

    def __str__(self) -> str:
        return f"{self.short_id} ({self.time})"


Snapshots = RootModel[list[Snapshot]]


class File(BaseModel):
    name: str
    file_type: str = Field(alias="type")
    path: Path
    uid: int
    gid: int
    size: int | None = None
    mode: int
    permissions: str
    mtime: datetime
    atime: datetime
    ctime: datetime

    def __str__(self) -> str:
        return f"[{self.file_type}] {self.name} ({self.path}): {self.size or '0'} bytes"
