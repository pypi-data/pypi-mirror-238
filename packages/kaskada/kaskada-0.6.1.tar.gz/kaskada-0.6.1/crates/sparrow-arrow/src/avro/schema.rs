use super::Error;
use arrow::datatypes::{DataType, Field, IntervalUnit, Schema, TimeUnit};
use avro_schema::schema::{
    Field as AvroField, Fixed, FixedLogical, IntLogical, LongLogical, Record, Schema as AvroSchema,
};
use error_stack::ResultExt;
use itertools::Itertools;
use std::collections::HashMap;

#[derive(derive_more::Display, Debug)]
#[display(fmt = "in field '{_0}'")]
struct FieldName(String);

/// Converts an arrow [`Schema`] to an Avro [`AvroSchema`].
pub fn to_avro_schema(schema: &Schema) -> error_stack::Result<AvroSchema, Error> {
    to_avro_record_schema(schema).map(AvroSchema::Record)
}

/// Converts an arrow [`Schema`] to an Avro [`Record`].
fn to_avro_record_schema(schema: &Schema) -> error_stack::Result<Record, Error> {
    let mut name_counter: i32 = 0;
    let fields: Vec<_> = schema
        .fields
        .iter()
        .map(|f| arrow_to_avro(f, &mut name_counter))
        .try_collect()?;

    Ok(Record {
        name: "Record".to_string(),
        namespace: None,
        doc: None,
        aliases: vec![],
        fields,
    })
}

fn arrow_to_avro(field: &Field, name_counter: &mut i32) -> error_stack::Result<AvroField, Error> {
    let schema = type_to_schema(field.data_type(), field.is_nullable(), name_counter)
        .attach_printable_lazy(|| FieldName(field.name().to_owned()))?;
    Ok(AvroField::new(field.name(), schema))
}

fn type_to_schema(
    data_type: &DataType,
    is_nullable: bool,
    name_counter: &mut i32,
) -> error_stack::Result<AvroSchema, Error> {
    // Nullable types are represented in Avro as the union of `null` and the type.
    Ok(if is_nullable {
        AvroSchema::Union(vec![
            AvroSchema::Null,
            type_to_schema_non_nullable(data_type, name_counter)?,
        ])
    } else {
        type_to_schema_non_nullable(data_type, name_counter)?
    })
}

fn record_name(name_counter: &mut i32) -> String {
    *name_counter += 1;
    format!("r{name_counter}")
}

fn type_to_schema_non_nullable(
    data_type: &DataType,
    name_counter: &mut i32,
) -> error_stack::Result<AvroSchema, Error> {
    Ok(match data_type {
        DataType::Null => AvroSchema::Null,
        DataType::Boolean => AvroSchema::Boolean,
        DataType::Int32 => AvroSchema::Int(None),
        DataType::Int64 => AvroSchema::Long(None),
        DataType::Float32 => AvroSchema::Float,
        DataType::Float64 => AvroSchema::Double,
        DataType::Binary => AvroSchema::Bytes(None),
        DataType::LargeBinary => AvroSchema::Bytes(None),
        DataType::Utf8 => AvroSchema::String(None),
        DataType::LargeUtf8 => AvroSchema::String(None),
        DataType::Struct(fields) => AvroSchema::Record(Record::new(
            record_name(name_counter),
            fields
                .iter()
                .map(|f| arrow_to_avro(f, name_counter))
                .try_collect()?,
        )),
        DataType::Date32 => AvroSchema::Int(Some(IntLogical::Date)),
        DataType::Time32(TimeUnit::Millisecond) => AvroSchema::Int(Some(IntLogical::Time)),
        DataType::Time64(TimeUnit::Microsecond) => AvroSchema::Long(Some(LongLogical::Time)),
        DataType::Timestamp(TimeUnit::Millisecond, None) => {
            AvroSchema::Long(Some(LongLogical::LocalTimestampMillis))
        }
        DataType::Timestamp(TimeUnit::Microsecond, None) => {
            AvroSchema::Long(Some(LongLogical::LocalTimestampMicros))
        }
        DataType::Interval(IntervalUnit::MonthDayNano) => {
            let mut fixed = Fixed::new("", 12);
            fixed.logical = Some(FixedLogical::Duration);
            AvroSchema::Fixed(fixed)
        }
        DataType::FixedSizeBinary(size) => AvroSchema::Fixed(Fixed::new("", (*size) as usize)),
        other => error_stack::bail!(Error::UnimplementedArrow(other.clone())),
    })
}

/// Converts an Avro [`AvroSchema`] to an arrow [`Schema`].
pub fn from_avro_schema(avro_schema: &AvroSchema) -> error_stack::Result<Schema, Error> {
    let mut names: HashMap<&str, usize> = HashMap::new();
    let fields = match avro_schema {
        AvroSchema::Record(record) => record
            .fields
            .iter()
            .map(|field| avro_to_arrow(field, &mut names))
            .collect::<Result<Vec<_>, _>>()?,
        _ => error_stack::bail!(Error::UnsupportedAvro(avro_schema.clone())),
    };
    Ok(Schema::new(fields))
}

fn avro_to_arrow(
    field: &AvroField,
    names: &mut HashMap<&str, usize>,
) -> error_stack::Result<Field, Error> {
    let name = field.name.as_str();
    let (data_type, nullable) = schema_to_type(&field.schema, names)?;
    Ok(Field::new(name, data_type, nullable))
}

/// returns (DataType, is_nullable)
fn schema_to_type(
    schema: &AvroSchema,
    names: &mut HashMap<&str, usize>,
) -> error_stack::Result<(DataType, bool), Error> {
    Ok(match schema {
        AvroSchema::Null => (DataType::Null, false),
        AvroSchema::Boolean => (DataType::Boolean, false),
        AvroSchema::Int(None) => (DataType::Int32, false),
        AvroSchema::Long(None) => (DataType::Int64, false),
        AvroSchema::Float => (DataType::Float32, false),
        AvroSchema::Double => (DataType::Float64, false),
        AvroSchema::Int(Some(IntLogical::Date)) => (DataType::Date32, false),
        AvroSchema::Int(Some(IntLogical::Time)) => (DataType::Time32(TimeUnit::Millisecond), false),
        AvroSchema::Long(Some(LongLogical::Time)) => {
            (DataType::Time64(TimeUnit::Microsecond), false)
        }
        AvroSchema::Long(Some(LongLogical::LocalTimestampMillis)) => {
            (DataType::Timestamp(TimeUnit::Millisecond, None), false)
        }
        AvroSchema::Long(Some(LongLogical::LocalTimestampMicros)) => {
            (DataType::Timestamp(TimeUnit::Microsecond, None), false)
        }
        AvroSchema::Bytes(None) => (DataType::Binary, false),
        AvroSchema::String(None) => (DataType::Utf8, false),
        AvroSchema::Fixed(fixed) => (DataType::FixedSizeBinary(fixed.size as i32), false),
        AvroSchema::Record(record) => (record_to_type(record, names)?, false),
        AvroSchema::Union(union) => {
            if union.len() == 2 && union.contains(&AvroSchema::Null) {
                let non_null_schema = union.iter().find(|&s| s != &AvroSchema::Null).unwrap();
                (schema_to_type(non_null_schema, names)?.0, true)
            } else {
                error_stack::bail!(Error::UnimplementedAvro(schema.clone()));
            }
        }
        _ => error_stack::bail!(Error::UnimplementedAvro(schema.clone())),
    })
}

fn record_to_type(
    record: &Record,
    names: &mut HashMap<&str, usize>,
) -> error_stack::Result<DataType, Error> {
    let fields = record
        .fields
        .iter()
        .map(|field| avro_to_arrow(field, names))
        .collect::<Result<Vec<_>, _>>()?;
    Ok(DataType::Struct(fields.into()))
}

#[cfg(test)]
mod tests {
    use super::to_avro_schema;
    use crate::avro::from_avro_schema;
    use arrow::datatypes::{DataType, Field, Schema};

    #[test]
    fn test_simple_struct_to_avro() {
        let arrow_schema = simple_arrow_schema();
        let avro_schema = to_avro_schema(&arrow_schema).unwrap();
        insta::assert_json_snapshot!(&avro_schema);
    }

    fn simple_arrow_schema() -> Schema {
        Schema::new(vec![
            Field::new("a", DataType::Int64, true),
            Field::new("b", DataType::Float64, false),
            Field::new("c", DataType::Utf8, true),
        ])
    }

    #[test]
    fn test_nested_struct_to_avro() {
        let arrow_schema = nested_arrow_schema();
        let avro_schema = to_avro_schema(&arrow_schema).unwrap();
        insta::assert_json_snapshot!(&avro_schema);
    }

    fn nested_arrow_schema() -> Schema {
        Schema::new(vec![
            Field::new("a", DataType::Int64, true),
            Field::new("b", DataType::Float64, false),
            Field::new("c", DataType::Utf8, true),
            Field::new(
                "d",
                DataType::Struct(
                    vec![
                        Field::new("a", DataType::Int64, true),
                        Field::new("b", DataType::Float64, false),
                    ]
                    .into(),
                ),
                false,
            ),
        ])
    }

    #[test]
    fn test_nullable_struct_to_avro() {
        let arrow_schema = nullable_arrow_schema();
        let avro_schema = to_avro_schema(&arrow_schema).unwrap();
        insta::assert_json_snapshot!(&avro_schema);
    }

    fn nullable_arrow_schema() -> Schema {
        Schema::new(vec![
            Field::new("a", DataType::Int64, true),
            Field::new(
                "b",
                DataType::Struct(
                    vec![
                        Field::new("c", DataType::Utf8, false),
                        Field::new("d", DataType::Int32, true),
                    ]
                    .into(),
                ),
                true,
            ),
        ])
    }

    fn test_round_trip(arrow_schema: Schema) {
        let avro_schema = to_avro_schema(&arrow_schema).unwrap();
        let result = from_avro_schema(&avro_schema).unwrap();
        assert_eq!(result, arrow_schema)
    }

    #[test]
    fn test_simple_round_trip() {
        test_round_trip(simple_arrow_schema());
    }

    #[test]
    fn test_nested_round_trip() {
        test_round_trip(nested_arrow_schema());
    }

    #[test]
    fn test_nullable_round_trip() {
        test_round_trip(nullable_arrow_schema());
    }
}
