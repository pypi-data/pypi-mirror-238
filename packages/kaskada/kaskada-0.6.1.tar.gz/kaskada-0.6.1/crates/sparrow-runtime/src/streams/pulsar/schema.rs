use arrow::array::ArrowPrimitiveType;
use arrow::datatypes::{Schema, SchemaRef, TimestampMicrosecondType};

use error_stack::{IntoReport, Result, ResultExt};
use std::sync::Arc;

#[derive(Debug, derive_more::Display)]
pub enum Error {
    #[allow(dead_code)]
    AvroNotEnabled,
    AvroSchemaConversion,
    SchemaRequest,
    UnsupportedSchema,
}
impl error_stack::Context for Error {}

pub fn get_output_schema(schema: SchemaRef) -> Result<SchemaRef, Error> {
    let fields = schema.fields();

    // Avro does not support certain types that we use internally for the implicit columns.
    // For the `_time` (timestamp_ns) column, we cast to timestamp_us, sacrificing nano precision.
    // (Optionally, we could provide a separate column composed of the nanos as a separate i64).
    // The `_subsort` and `_key_hash` columns are dropped.
    let time_us = Arc::new(arrow::datatypes::Field::new(
        "_time",
        TimestampMicrosecondType::DATA_TYPE,
        false,
    ));

    let mut new_fields = vec![time_us];
    fields
        .iter()
        .skip(3) // Skip the `_time`, `_subsort`, and `_key_hash` fields
        .for_each(|f| new_fields.push(f.clone()));

    Ok(Arc::new(Schema::new(new_fields)))
}

#[cfg(not(feature = "avro"))]
fn format_schema(_schema: SchemaRef) -> Result<String, Error> {
    error_stack::bail!(Error::AvroNotEnabled)
}

#[cfg(feature = "avro")]
pub fn format_schema(schema: SchemaRef) -> Result<String, Error> {
    let avro_schema = sparrow_arrow::avro::to_avro_schema(&schema)
        .change_context(Error::AvroSchemaConversion)
        .attach_printable_lazy(|| {
            format!("failed to convert arrow schema to avro schema {schema}")
        })?;
    serde_json::to_string(&avro_schema)
        .into_report()
        .change_context(Error::AvroSchemaConversion)
        .attach_printable_lazy(|| {
            format!("failed to serialize avro schema to json string: {avro_schema:?}")
        })
}

fn schema_from_formatted(formatted_schema: &str) -> Result<Schema, Error> {
    // construct an avro_schema::schema::Schema from the json formatted schema
    let avro_schema: avro_schema::schema::Schema = serde_json::from_str(formatted_schema)
        .into_report()
        .change_context(Error::AvroSchemaConversion)
        .attach_printable_lazy(|| {
            format!("failed to deserialize avro schema from json string: {formatted_schema}")
        })?;
    // convert to sparrow format
    let sparrow_schema = sparrow_arrow::avro::from_avro_schema(&avro_schema)
        .change_context(Error::AvroSchemaConversion)
        .attach_printable_lazy(|| format!("from_avro_schema({:?}) failed", &avro_schema))?;
    Ok(sparrow_schema)
}

#[derive(serde::Deserialize, Debug)]
struct SchemaResponse {
    #[allow(dead_code)]
    version: u32,
    #[serde(rename = "type")]
    schema_type: String,
    #[allow(dead_code)]
    timestamp: u64,
    data: String,
    #[allow(dead_code)]
    properties: serde_json::Value,
}

// auth_params looks like
// token:xxx
pub fn pulsar_auth_token(auth_params: &str) -> Result<&str, Error> {
    // split on the first colon
    let mut parts = auth_params.splitn(2, ':');
    // verify first part is "token"
    let auth_type = parts.next();
    let Some(auth_type) = auth_type else {
        return Err(Error::SchemaRequest.into()).attach_printable("missing auth_type");
    };
    if auth_type != "token" {
        return Err(Error::SchemaRequest.into())
            .attach_printable_lazy(|| format!("auth type {:?}", auth_type));
    }
    let auth_token = parts.next();
    let Some(auth_token) = auth_token else {
        return Err(Error::SchemaRequest.into()).attach_printable("missing auth token");
    };
    Ok(auth_token)
}

// retrieve the schema for the given topic via the admin api, with a REST call.
// we can't use the pulsar client because the schema is not exposed there in the Rust client.
pub async fn get_pulsar_schema(
    admin_service_url: &str,
    tenant: &str,
    namespace: &str,
    topic: &str,
    auth_params: &str,
) -> Result<Schema, Error> {
    let url = format!(
        "{}/admin/v2/schemas/{}/{}/{}/schema",
        admin_service_url, tenant, namespace, topic
    );
    tracing::debug!("requesting schema from {}", url);
    let client = reqwest::Client::new();
    let mut request_builder = client.get(&url);

    // Auth is generally recommended, but some local builds may be run
    // without auth for exploration/testing purposes.
    if !auth_params.is_empty() {
        request_builder = request_builder.header(
            reqwest::header::AUTHORIZATION,
            format!("Bearer {}", pulsar_auth_token(auth_params)?),
        )
    }

    let text = request_builder
        .send()
        .await
        .into_report()
        .change_context(Error::SchemaRequest)?
        .text()
        .await
        .into_report()
        .change_context(Error::SchemaRequest)?;

    let schema_response: SchemaResponse = serde_json::from_str(&text)
        .into_report()
        .change_context(Error::AvroSchemaConversion)
        .attach_printable_lazy(|| format!("from_str({:?}) failed", &text))?;
    if schema_response.schema_type != "AVRO" {
        return Err(Error::UnsupportedSchema.into())
            .attach_printable_lazy(|| format!("schema type {:?}", schema_response.schema_type));
    }
    let schema = schema_from_formatted(&schema_response.data)?;

    Ok(schema)
}
