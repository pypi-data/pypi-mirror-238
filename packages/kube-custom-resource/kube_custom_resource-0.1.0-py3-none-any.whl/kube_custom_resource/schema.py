import copy
import enum
import typing

from pydantic import (
    BaseModel as PydanticModel,
    Extra,
    ValidationError,
    ConstrainedInt,
    ConstrainedFloat,
)


def resolve_refs(schema, definitions):
    """
    Recursively resolve $refs in the given schema using the definitions.
    """
    if isinstance(schema, dict):
        if "allOf" in schema and len(schema["allOf"]) == 1:
            # Where the schema has an allOf with a single item, just put the
            # fields from the item onto the schema
            items = schema.pop("allOf")[0]
            resolve_refs(items, definitions)
            schema.update(items)
        elif "$ref" in schema:
            ref = schema.pop("$ref").removeprefix("#/definitions/")
            referenced = definitions[ref]
            resolve_refs(referenced, definitions)
            schema.update(definitions[ref])
        else:
            for value in schema.values():
                resolve_refs(value, definitions)
    elif isinstance(schema, list):
        for item in schema:
            resolve_refs(item, definitions)


def remove_fields(schema, *fields):
    """
    Recursively remove the specified fields from all the types in the schema.
    """
    if isinstance(schema, dict):
        if "type" in schema:
            for field in fields:
                schema.pop(field, None)
        for item in schema.values():
            remove_fields(item, *fields)
    elif isinstance(schema, list):
        for item in schema:
            remove_fields(item, *fields)


def snake_to_pascal(name):
    """
    Converts a snake case name to pascalCase.
    """
    first, *rest = name.split("_")
    return "".join([first] + [part.capitalize() for part in rest])


class Enum(enum.Enum):
    """
    Enum that does not include a title in the JSON-Schema.
    """
    def __str__(self):
        return str(self.value)

    @classmethod
    def __modify_schema__(cls, field_schema):
        field_schema.pop("title", None)


class Any:
    """
    Type for a value that can be any type.
    """
    @classmethod
    def __modify_schema__(cls, field_schema):
        field_schema["x-kubernetes-preserve-unknown-fields"] = True

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        return v


class Dict(typing.Dict):
    """
    Dict whose JSON-Schema includes the custom attribute to prevent Kubernetes
    pruning unknown properties.
    """
    @classmethod
    def __modify_schema__(cls, field_schema):
        field_schema["x-kubernetes-preserve-unknown-fields"] = True


class IntOrString(str):
    """
    Type for a value that can be specified as an integer or a string.

    The value will resolve to a string and the generated schema will include the
    Kubernetes custom schema attribute 'x-kubernetes-int-or-string'.
    """
    @classmethod
    def __modify_schema__(cls, field_schema):
        field_schema.pop("type", None)
        field_schema.update({
            "x-kubernetes-int-or-string": True,
            "anyOf": [
                { "type": "integer" },
                { "type": "string" },
            ],
        })

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        if not isinstance(v, (str, int)):
            raise TypeError("int or string required")
        return str(v)


class ConstrainedNumberMixin:
    @classmethod
    def __modify_schema__(cls, field_schema: typing.Dict[str, typing.Any]) -> None:
        super().__modify_schema__(field_schema)
        exclusive_min = field_schema.pop("exclusiveMinimum", None)
        if exclusive_min is not None:
            field_schema.update({
                "minimum": exclusive_min,
                "exclusiveMinimum": True,
            })
        exclusive_max = field_schema.pop("exclusiveMaximum", None)
        if exclusive_max is not None:
            field_schema.update({
                "maximum": exclusive_max,
                "exclusiveMaximum": True,
            })


def conint(
    *,
    strict: bool = False,
    gt: int = None,
    ge: int = None,
    lt: int = None,
    le: int = None,
    multiple_of: int = None
) -> typing.Type[int]:
    return type(
        "ConstrainedIntValue",
        (ConstrainedNumberMixin, ConstrainedInt),
        dict(
            strict = strict,
            gt = gt,
            ge = ge,
            lt = lt,
            le = le,
            multiple_of = multiple_of
        )
    )


def confloat(
    *,
    strict: bool = False,
    gt: float = None,
    ge: float = None,
    lt: float = None,
    le: float = None,
    multiple_of: float = None,
    allow_inf_nan: typing.Optional[bool] = None,
) -> typing.Type[float]:
    # use kwargs then define conf in a dict to aid with IDE type hinting
    return type(
        "ConstrainedFloatValue",
        (ConstrainedNumberMixin, ConstrainedFloat),
        dict(
            strict = strict,
            gt = gt,
            ge = ge,
            lt = lt,
            le = le,
            multiple_of = multiple_of,
            allow_inf_nan = allow_inf_nan
        )
    )


class StructuralUnion:
    """
    Type for a structural union, i.e. a union with a structural schema.

    See https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema.
    """
    def __class_getitem__(cls, types):
        name = f"{cls.__name__}[{','.join(t.__name__ for t in types)}]"
        return type(name, (cls, ), {}, __types__ = types)

    def __init_subclass__(cls, /, __types__, **kwargs):
        # Structural unions are only supported for schema models
        if not all(issubclass(t, BaseModel) for t in __types__):
            raise TypeError("structural unions are only supported between schema models")
        super().__init_subclass__(**kwargs)
        cls.__types__ = __types__
    
    @classmethod
    def __modify_schema__(cls, field_schema):
        field_schema["type"] = "object"
        field_schema["properties"] = {}
        if cls.__doc__:
            field_schema["description"] = cls.__doc__
        field_schema["anyOf"] = []
        for type in cls.__types__:
            # The schema is cached, so make sure to copy it before modifying it
            schema = copy.deepcopy(type.schema())
            # In order to qualify as a structural schema, the schema of the union itself
            # must include all the possible properties
            field_schema["properties"].update(copy.deepcopy(schema["properties"]))
            # Schemas in anyOf are not permitted to contain particular keys
            remove_fields(
                schema,
                "description",
                "type",
                "default",
                "additionalProperties",
                "nullable",
                "x-kubernetes-preserve-unknown-fields",
            )
            field_schema["anyOf"].append(schema)

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        # Try to instantiate each of the types in the union
        # The first one to succeed is used
        for type in cls.__types__:
            if isinstance(v, type):
                return v
            elif isinstance(v, typing.Mapping):
                try:
                    return type(**v)
                except (ValueError, TypeError, ValidationError):
                    pass
        raise TypeError("failed to select a type from the union")


class BaseModel(PydanticModel):
    """
    Base model for use within CRD definitions.
    """
    class Config:
        # Allow pascalCase names as well as snake_case
        alias_generator = snake_to_pascal
        allow_population_by_field_name = True
        # Validate any mutations to the model
        allow_mutation = True
        validate_assignment = True

        @classmethod
        def schema_extra(cls, schema, model):
            """
            Post-process the generated schema to make it compatible with a Kubernetes CRD.
            """
            # Remove the titles
            schema.pop("title", None)
            for prop in schema.get("properties", {}).values():
                prop.pop("title", None)
            # When extra fields are allowed, stop Kubernetes pruning them
            if model.__config__.extra == Extra.allow:
                schema["x-kubernetes-preserve-unknown-fields"] = True

    def dict(self, **kwargs):
        # Unless otherwise specified, we want by_alias = True
        kwargs.setdefault("by_alias", True)
        return super().dict(**kwargs)

    def json(self, **kwargs):
        # Unless otherwise specified, we want by_alias = True
        kwargs.setdefault("by_alias", True)
        return super().json(**kwargs)

    @classmethod
    def schema(cls, *args, include_defaults = False, **kwargs):
        schema = super().schema(*args, **kwargs)
        # If the schema has definitions defined, resolve $refs and remove them
        if "definitions" in schema:
            resolve_refs(schema, schema.pop("definitions"))
        # Unless explicitly included, we remove defaults from the schema as they cause
        # Kubernetes to rewrite the schema
        # In most cases, it is better that defaults are applied at model instantiation time
        # as rewriting the Kubernetes objects themselves can have unintended side-effects
        # However in some cases it is more appropriate for the defaults to be "locked in" at
        #Â creation time
        if not include_defaults:
            remove_fields(schema, "default")
        return schema
