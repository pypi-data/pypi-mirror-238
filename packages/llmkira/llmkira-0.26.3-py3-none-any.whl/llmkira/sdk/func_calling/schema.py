# -*- coding: utf-8 -*-

import re
from abc import abstractmethod, ABC
from dataclasses import dataclass, field
from types import ModuleType
from typing import Optional, Type, Dict, Any, List, Union, Set, final, Literal

from loguru import logger
from pydantic import BaseModel, Field, validator, root_validator

from ..schema import Function


class BaseTool(ABC, BaseModel):
    """
    基础工具类，所有工具类都应该继承此类
    """
    __slots__ = ()
    silent: bool = Field(False, description="是否静默")
    function: Function = Field(..., description="功能")
    keywords: List[str] = Field([], description="关键词")
    pattern: Optional[re.Pattern] = Field(None, description="正则匹配")
    require_auth: bool = Field(False, description="是否需要授权")
    repeatable: bool = Field(False, description="是否可重复使用")
    deploy_child: Literal[0, 1] = Field(1, description="如果为0，终结于此链点，不再向下传递")
    require_auth_kwargs: dict = {}
    env_required: List[str] = Field([], description="环境变量要求")
    file_match_required: Optional[re.Pattern] = Field(None, description="re.compile 文件名正则")

    # exp: re.compile(r"file_id=([a-z0-9]{8})")

    @final
    @property
    def name(self):
        """
        工具名称
        """
        return self.function.name

    @final
    @root_validator
    def _check_conflict(cls, values):
        # env_required and silent
        if values["silent"] and values["env_required"]:
            raise ValueError("silent and env_required can not be True at the same time")
        return values

    @final
    @validator("keywords", pre=True)
    def _check_keywords(cls, v):
        for i in v:
            if not isinstance(i, str):
                raise ValueError(f"keyword must be str, got {type(i)}")
            if len(i) > 20:
                raise ValueError(f"keyword must be less than 20 characters, got {len(i)}")
            if len(i) < 2:
                raise ValueError(f"keyword must be more than 2 characters, got {len(i)}")
        return v

    def env_help_docs(self, empty_env: List[str]) -> str:
        """
        环境变量帮助文档
        :param empty_env: 未被配置的环境变量列表
        :return: 帮助文档/警告
        """
        assert isinstance(empty_env, list), "empty_env must be list"
        return "You need to configure ENV to start use this tool"

    @abstractmethod
    def pre_check(self) -> Union[bool, str]:
        """
        预检查，如果不合格则返回 False，合格则返回 True
        返回字符串表示不合格，且有原因
        """
        return ...

    @abstractmethod
    def func_message(self, message_text, **kwargs):
        """
        如果合格则返回message，否则返回None，表示不处理
        """
        for i in self.keywords:
            if i in message_text:
                return self.function
        # 正则匹配
        if self.pattern:
            match = self.pattern.match(message_text)
            if match:
                return self.function
        return None

    @abstractmethod
    async def failed(self, task, receiver, arg, exception, **kwargs):
        """
        处理失败
        """
        return ...

    @abstractmethod
    async def callback(self, task, receiver, arg, result, **kwargs):
        """
        成功回调
        """
        return ...

    @abstractmethod
    async def run(self, task, receiver, arg, **kwargs):
        """
        处理message，返回message
        """
        env = kwargs.get("env", {})
        return ...

    @final
    async def load(self, task, receiver, arg, env: dict = None, **kwargs):
        if not env:
            env = {}
        run_arg = {
            "task": task,
            "receiver": receiver,
            "arg": arg,
            "env": env
        }
        try:
            run_result = await self.run(
                **run_arg
            )
        except Exception as e:
            logger.error(f"ToolExecutor:run error: {e}")
            run_arg["exception"] = e
            await self.failed(
                **run_arg
            )
        else:
            run_arg["result"] = run_result
            try:
                await self.callback(
                    **run_arg
                )
            except Exception as e:
                logger.error(f"ToolExecutor:callback error: {e}")
        return run_arg


@dataclass(eq=False)
class FuncPair:
    function: Function
    tool: Type[BaseTool]
    extra_arg: Dict[Any, Any] = field(default_factory=dict)

    @property
    def name(self):
        return self.function.name


@dataclass(eq=False)
class PluginMetadata:
    """
    插件元信息，由插件编写者提供
    """

    name: str
    """插件名称"""
    description: str
    """插件功能介绍"""
    usage: str
    """插件使用方法"""
    openapi_version: str
    """适应版本"""
    function: Set[FuncPair] = field(default_factory=set)
    """插件工具"""
    type: Optional[str] = None
    """插件类型"""
    homepage: Optional[str] = None
    """插件主页"""
    config: Optional[Type[BaseModel]] = None
    """插件配置项"""
    extra: Dict[Any, Any] = field(default_factory=dict)
    """插件额外信息，可由插件编写者自由扩展定义"""


@dataclass(eq=False)
class Plugin:
    """
    机制用途，不由用户实例化
    """

    name: str
    """插件索引标识，Bot 使用 文件/文件夹 名称作为标识符"""
    module: ModuleType
    """插件模块对象"""
    module_name: str
    """点分割模块路径"""
    manager: "PluginManager"
    """导入该插件的插件管理器"""
    parent_plugin: Optional["Plugin"] = None
    """父插件"""
    sub_plugins: Set["Plugin"] = field(default_factory=set)
    """子插件集合"""
    metadata: Optional[PluginMetadata] = None
    """插件元信息"""

# 探针插件加载机制 参考 nonebot2 项目设计
