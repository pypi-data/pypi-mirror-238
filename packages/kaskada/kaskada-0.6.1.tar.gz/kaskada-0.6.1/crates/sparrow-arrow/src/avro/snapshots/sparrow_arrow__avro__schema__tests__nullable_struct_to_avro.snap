---
source: crates/sparrow-arrow/src/avro/schema.rs
expression: "&avro_schema"
---
{
  "type": "record",
  "name": "Record",
  "fields": [
    {
      "name": "a",
      "type": [
        "null",
        "long"
      ]
    },
    {
      "name": "b",
      "type": [
        "null",
        {
          "type": "record",
          "name": "r1",
          "fields": [
            {
              "name": "c",
              "type": "string"
            },
            {
              "name": "d",
              "type": [
                "null",
                "int"
              ]
            }
          ]
        }
      ]
    }
  ]
}
